enhancements/20221228-various-storage-class-support.md

```diff
@@ -0,0 +1,114 @@
+# Various StorageClass support of Harvester CSI Driver
+
+## Summary
+
+Harvester supports the user-defined StorageClass after v1.1.0. Because of that, Harvester CSI Driver should also let users use the custom StorageClass mentioned above.
+
+### Related Issues
+
+https://github.com/harvester/harvester/issues/3052
+
+## Motivation
+
+### Goals
+
+The user using rancher integration with Harvester Cluster (Called Guest Cluster) could create the volume with the different StorageClass associated with StorageClass on the Host Harvester Cluster.
+
+## Proposal
+
+We will implement the mechanism to get the StorageClass parameter from the Guest Cluster and try to create it on the Host Harvester Cluster.
+
+### User Stories
+
+#### General use cases
+
+Generally, users use the default StorageClass(`harvester`). The default StorageClass has three replicas means higher reliability. Users do not do any extra configuration. Create volume with the default StorageClass(`harvester`), then use.
+
+#### More specific goals
+
+Sometimes users will have different purposes for storage volume. Like `data archiving` or `Storage tiering`. For the above requirements, users can change the data locality to HDD. Users can complete it by creating a new StorageClass and configuring its data locality to HDD. Also, for the `performance` purpose. Users can reduce the number of replicas (remember, the default replica number is three). Reducing the replica number from 3 to 2 would increase the volume performance, but it lost some data reliability. So, users can create a new StorageClass and config its replica number to 2 and make the storage volume with this StorageClass to get better performance. 
```

Comment:
- Is changing the data locality to HDD considered part a change to increase performance? If not, it might be more clear if "Also, for the `performance` purpose. Users..." is changed to "For performance purposes, users...".
- Reducing the replicas from 3 to 2 is called out, is 2 the minimum or is going from 3 to 1 a possibility?
- `but it lost some data reliability` -> `but results in some lost data reliability`

Reply:
> * Is changing the data locality to HDD considered part a change to increase performance? If not, it might be more clear if "Also, for the `performance` purpose. Users..." is changed to "For performance purposes, users...".

No, they are different purposes. I will change to "For performance purposes...". Thanks!

> * Reducing the replicas from 3 to 2 is called out, is 2 the minimum or is going from 3 to 1 a possibility?
I thought that reducing the replica to 1 was too aggressive. So I do not want to write it down. If we only focus on performance, 1 replica would get the better performance with write IO.

> * `but it lost some data reliability` -> `but results in some lost data reliability`

OK!

Reply:
> I thought that reducing the replica to 1 was too aggressive. So I do not want to write it down. If we only focus on performance, 1 replica would get the better performance with write IO. WDUT?

Not explicitly mentioning 3 -> 1 sounds good to me, but maybe just a slight adjustment to indicate 3 -> 2 is an example and not the only choice.

E.g. `Reducing the replica number increases the volume's performance, but results in some lost data reliability. For example, users can create a new StorageClass and configure its replica number to 2 and make the storage volume with this StorageClass to get better performance.`.

Reply:
Done!  Thanks for the review helping.